sycl/doc/design/KernelFusionJIT.md

```diff
@@ -0,0 +1,173 @@
+
+The pipeline currently consists of the following passes (in order):
+
+  - `SYCLKernelFusion` performs the actual fusion process by inlining kernels to fuse inside the fused kernel
+  - Generic optimization passes: `IndVarSimplifyPass`, `LoopUnrollPass`, `SROAPass`, `InferAddressSpacesPass` to remove pointers to the generic address-space
+    - These optimizations are important to help the internalizer, see note below.
+  - `SYCLInternalizer` promotes buffer to local or private memory
+  - `SYCLCP` propagates runtime information as constant in the kernel
+  - Generic optimization passes post-fusion: `SROAPass`, `SCCPPass`, `InstCombinePass`, `SimplifyCFGPass`, `SROAPass`, `InstCombinePass`, `SimplifyCFGPass`, `ADCEPass`, `EarlyCSEPass`
+
+Note: ideally the `InferAddressSpacesPass` task should be done by the static compiler.
+However, to help the inference and (in fine) the internalizer passes in general, we need to run optimizations more aggressively than the static currently does.
+The main barrier that could prevent the internalization of buffers is the presence of generic pointers.
+As we need to change a pointer to generic to a pointer to private / local, if it is casted to generic we need to ensure that any casts back
+to global can be safely changed to the proper address space.
+A more precise inference pass and memory analysis will be required to lift this workaround.
+
+Once the pipeline is finished, the module is translated into a SPIR-V module and encapsulated into a `SYCLKernelInfo` object containing metadata required for its injection into the runtime.
+
+### Passing information to the fusion pipeline
+
+Most of the information passed from the runtime into the JIT is stored as metadata inside the LLVM module the JIT creates.
+This eases lit testing as all the information is self-contained in the module.
+The metadata is attached to a function that will become the fused kernel:
+
+- `sycl.kernel.fused`: declare the kernels to fuse. Contains a list of kernel names to fuse inside the function
+- `sycl.kernel.param`: declare identical parameters. Contains a list of tuples, each tuple represents identical arguments and each element of that tuple contains a pair of indexes referencing the kernel index in `sycl.kernel.fused` and the parameter index of that kernel (0 indexed). For instance ((0,1),(2,3)) means the second argument of the first kernel is identical to the fourth argument of the third kernel.
+- `sycl.kernel.promote`: declare identical parameters to be promoted. Contains a list of index (of the fused kernel, after identical arguments elision) and `private` if the argument is to be promoted to private memory or `local` if it is to local.
+- `sycl.kernel.promote.localsize`: declare the address space size for the promoted memory. Contains a list of indexes (of the fused kernel, after identical arguments elision) and the number of elements.
```

Comment:
Does it include only `local` memory or all promoted memory? The name of the metadata does not correlate with the description very well.

Reply:
Yes indeed, legacy of the first prototype. I changed to `sycl.kernel.promote.size` which is way more neutral.